File: brain/vtep/src/main/scala/org/midonet/vtep/OvsdbTranslator.scala
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.vtep

import java.util

import scala.collection.JavaConversions.{asScalaSet, mapAsJavaMap, mapAsScalaMap, setAsJavaSet}

import org.opendaylight.ovsdb.lib.notation.{UUID => OvsdbUUID}

import org.midonet.packets.IPv4Addr

/**
 * Type translations to and from ovsdb-specific data models
 */
object OvsdbTranslator {

    /** Convert from Ovsdb UUIDs to Java UUIDs */
    def fromOvsdb(id: OvsdbUUID): util.UUID = util.UUID.fromString(id.toString)

    /** Convert from Java UUIDs to Ovsdb UUIDs */
    def toOvsdb(uuid: util.UUID): OvsdbUUID = new OvsdbUUID(uuid.toString)

    /** Convert a set of OvsdbUUIDs to a set of Java UUIDs
      * The input set can be a null value, resulting in an empty * output set */
    def fromOvsdb(inSet: util.Set[_]): util.Set[util.UUID] =
        if (inSet == null) new util.HashSet[util.UUID]()
        else setAsJavaSet(asScalaSet[OvsdbUUID](
            inSet.asInstanceOf[util.Set[OvsdbUUID]]).map(fromOvsdb))

    /** Convert a set of Java UUIDs to a set of ovsdb UUIDs
      * The input set can be a null value, resulting in an empty output set */
    def toOvsdb(inSet: Set[util.UUID]): util.Set[OvsdbUUID] =
        if (inSet == null) new util.HashSet[OvsdbUUID]()
        else setAsJavaSet(inSet.map(toOvsdb))

    /** Convert from (Long -> OvsdbUUID) to a map of (Integer -> java UUID)
     * The input map can be a null value, resulting in an empty output set */
    def fromOvsdb(inMap: util.Map[_, _]): util.Map[Integer, util.UUID] =
        if (inMap == null) new util.HashMap[Integer, util.UUID]()
        else mapAsJavaMap(
            mapAsScalaMap(inMap.asInstanceOf[util.Map[Long, OvsdbUUID]])
                .map(e => (e._1.toInt.asInstanceOf[Integer], fromOvsdb(e._2)))
                .toMap
        )

    /** Convert from (Integer -> java UUID) to a map of (Long -> OvsdbUUID)
      * The input map can be a null value, resulting in an empty output map */
    def toOvsdb(inMap: Map[Integer, util.UUID]): util.Map[Long, OvsdbUUID] =
        if (inMap == null) new util.HashMap[Long, OvsdbUUID]()
        else mapAsJavaMap(inMap.map(e => (e._1.toLong, toOvsdb(e._2))))

    /** Convert from a set of strings to a set of the ip addresses */
    def fromOvsdbIpSet(inSet: util.Set[_]): util.Set[IPv4Addr] =
        if (inSet == null) new util.HashSet[IPv4Addr]()
        else setAsJavaSet(
            asScalaSet(inSet.asInstanceOf[util.Set[String]]).collect({
                case s: String if s.nonEmpty => IPv4Addr.fromString(s)
            })
        )

    /** Convert from a set of IP addresses to a set of ovsdb strings */
    def toOvsdbIpSet(inSet: Set[IPv4Addr]): util.Set[String] =
        if (inSet == null) new util.HashSet[String]()
        else setAsJavaSet(inSet.collect{case s: IPv4Addr => s.toString})
}

File: brain/vtep/src/main/java/org/midonet/vtep/OvsdbTranslator.java
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.vtep;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;


import org.opendaylight.ovsdb.lib.notation.UUID;

import org.midonet.packets.IPv4Addr;

/**
 * Type translations to and from ovsdb-specific data models
 */
public class OvsdbTranslator {

    /**
     * Convert from Ovsdb UUIDs to Java UUIDs
     */
    static public java.util.UUID fromOvsdb(UUID id) {
        return java.util.UUID.fromString(id.toString());
    }

    /**
     * Convert from Java UUIDs to Ovsdb UUIDs
     */
    static public UUID toOvsdb(java.util.UUID uuid) {
        return new UUID(uuid.toString());
    }

    /**
     * Convert to a set of java UUIDs
     * The input set can be a null value, which will result in an empty
     * output set
     */
    static public Set<java.util.UUID> fromOvsdb(Set inSet) {
        Set<java.util.UUID> set = new HashSet<>();
        if (inSet != null) {
            for (Object id: inSet) {
                set.add(fromOvsdb((UUID)id));
            }
        }
        return set;
    }

    /**
     * Convert to a set of ovsdb UUIDs
     * The input set can be a null value, which will result in an empty
     * output set
     */
    static public Set<UUID> toOvsdb(Set<java.util.UUID> inSet) {
        Set<UUID> set = new HashSet<>();
        if (inSet != null) {
            for (java.util.UUID id: inSet) {
                set.add(toOvsdb(id));
            }
        }
        return set;
    }

    /**
     * Convert to a map of Integer - java UUIDs
     * The input map can be a null value, which will result in an empty
     * output set
     */
    static public Map<Integer, java.util.UUID> fromOvsdb(Map inMap) {
        Map<Integer, java.util.UUID> map = new HashMap<>();
        if (inMap != null) {
            for (Object k: inMap.keySet()) {
                map.put(((Long)k).intValue(), fromOvsdb((UUID)inMap.get(k)));
            }
        }
        return map;
    }

    /**
     * Convert from a set of strings to a set of the ip addresses
     */
    static public Set<IPv4Addr> fromOvsdbIpSet(Set inSet) {
        Set<IPv4Addr> set = new HashSet<>();
        if (inSet != null) {
            for (Object obj : inSet) {
                String str = (String)obj;
                if (!str.isEmpty())
                    set.add(IPv4Addr.fromString(str));
            }
        }
        return set;
    }
}
